in-progress/7228-unified-bbjs-interface.md

```diff
@@ -0,0 +1,247 @@
+
+:question: Can we reimplement this logic on ts-land? We've received conflicting responses to this.
+
+Native implementation (does not support `ClientIVC, AVM`)
+
+- `bb proof_as_fields | proof_as_fields_honk` (assumes `_honk` supports both Ultra and Mega)
+
+Wasm implementation (does not support `ClientIVC, AVM, MegaHonk`)
+
+- `srsInitSrs(Crs(1))`
+- `acirNewAcirComposer(0)`
+- `acirSerializeProofIntoFields | acirProofAsFieldsUltraHonk`
+
+#### `getVerificationKey(circuit: Circuit, flavor: Flavor): { vk: Buffer, vkAsFields: Fields }`
+
+Computes the verification key for a given circuit. Returns the verification key in binary and as fields, which allows the client to read its hash and the number of public inputs for the circuit. Users should call this once per circuit they want to verify, and reuse it for verifying as many proofs as needed.
+
+:question: Can we avoid returning `vkAsFields`, and convert from binary to fields in ts-land?
+
+Native implementation (`ClientIVC, AVM` not supported)
+
+- `bb write_vk | write_vk_ultra_honk | write_vk_mega_honk`
+- `bb vk_as_fields | vk_as_fields_ultra_honk | vk_as_fields_mega_honk`
+
+Wasm implementation (only `UltraPlonk` supported)
+
+- `acirGetCircuitSizes`
+- `commonInitSlabAllocator(size)`
+- `srsInitSrs(size)`
+- `acirInitProvingKey`
+- `acirGetVerificationKey`
+- `acirSerializeVerificationKeyIntoFields`
+
+#### `verify(proof: Buffer, vk: Buffer, flavor: Flavor): { isValid: boolean }`
+
+Verifies the given proof, which includes public inputs, against the given verification key. Both proof and vk are in raw binary format. Requires the output of `getVerificationKey` for the given circuit first.
+
+Native implementation (`ClientIVC` not supported)
+
+- `bb verify | avm_verify | verify_ultra_honk | verify_mega_honk`
+
+Wasm implementation (`ClientIVC, MegaHonk, AVM` not supported)
+
+- `srsInitSrs(Crs(1))`
+- `acirNewAcirComposer(0)`
+- `acirLoadVerificationKey`
+- `acirVerifyProof | acirVerifyUltraHonk`
+
+#### `getSolidityVerifier(vk: Buffer, flavor: Flavor): { contract: Buffer }`
+
+Generates a Solidity verifier contract for the given verification key in raw binary format. Only `UltraPlonk` is supported. Expected to be called once per circuit during development. Requires the output of `getVerificationKey`.
+
+Native implementation:
+
+- `bb contract`
+
+Wasm implementation:
+
+- `srsInitSrs(Crs(1))`
+- `acirNewAcirComposer(0)`
+- `acirLoadVerificationKey`
+- `acirGetSolidityVerifier`
+
+#### `getGateCount(circuit: Circuit, flavor: Flavor): { gates: number }`
```

Comment:
We should maintain the opcode count in the response very least.

The `gates` command also now returns the profiling data for the flamegraph, I'd be happy to push this off into a separate "profile" command or even the nonstandard interface for the time being.